frontend/src/components/utils.js

```diff
@@ -201,6 +201,85 @@ async function getSelfUser() {
   return user;
 }
 
+/**
+ * Make a mutable copy of a case item, and run a callback to mutate the new copy
+ * @param {*} caseItem root case item
+ * @param {(caseItem: any, type: string) => void} callback function to run at every step of the tree
+ * @returns
+ */
+function visitCaseItem(caseItem, callback, itemType = "TopLevelNormativeGoal") {
+  if (typeof caseItem != "object") {
+    return caseItem;
+  }
+
+  // make a shallow copy
+  var copy = { ...caseItem };
+
+  configData.navigation[itemType]["children"].forEach((childName, j) => {
+    let childType = configData.navigation[itemType]["children"][j];
+    let dbName = configData.navigation[childName]["db_name"];
+    // recurse to make deep copies of the child arrays, if they exist
+    if (Array.isArray(copy[dbName])) {
+      copy[dbName] = copy[dbName].map((g) =>
+        visitCaseItem(g, callback, childType),
+      );
+    }
+  });
+
+  callback(copy, itemType);
+
+  return copy;
+}
+
+/**
+ * For an assurance case, and a case item id and type,
+ * find all property claims this item in the graph,
+ * including itself if its a property claim
+ * @param {*} assuranceCase
+ * @param {string} id
+ * @param {string} type
+ * @returns {any[]}
+ */
+function getParentPropertyClaims(assuranceCase, id, type) {
+  // run depth first search
+  // because evidence cannot have property claim children,
+  // don't worry that they can appear twice in the tree
+  /** @type [any, string, any[]] */
+  const caseItemStack = assuranceCase.goals.map((i) => [
+    i,
+    "TopLevelNormativeGoal",
+    [],
+  ]);
+
+  let result = [];
+
+  while (caseItemStack.length > 0) {
+    const [node, nodeType, parents] = caseItemStack.shift();
+    const newParents = [...parents, node];
+
+    if (node.id.toString() === id && nodeType === type) {
+      // found it!
+      result = newParents;
+      break;
```

Comment:
Does this work if there's an Evidence item that has two Property Claim parents that are under different Top Level Goals? Say there's G1 -> P1 -> E1 and G2 -> P2 -> P2.1 -> E2. I didn't think this through carefully, but I worry that this loop might go G1 -> P1 -> E1 and then break, and miss out on P2 and P2.1.

Reply:
It would miss out on P2 and P2.1, yeah. For how we use it that's fine, as this function is only used for other property claims and not evidences, so maybe a name change is in order...

Reply:
Okay, thanks. If you feel like renaming/fixing feel free, to but it's minor enough, so I'll approve now.